packages/nimble/R/BUGS_modelDef.R

```diff
@@ -199,6 +200,58 @@ codeProcessIfThenElse <- function(code, constants, envir = parent.frame()) {
         return(code)
 }
 
+## This function recurses through a block of code and expands any submodels
+codeProcessModelMacros <- function(code, envir = parent.frame()) {
+    codeLength <- length(code)
+    ## First check if this is the start of a curly-bracketed block
+    if(code[[1]] == '{') {
+        if(codeLength > 1)
+            ## Recurse on each line
+            for(i in 2:codeLength)
+                code[[i]] <- codeProcessModelMacros(code[[i]], envir)
+        return(code)
+    }
+    ## If this is a for loop, recurse on the body of the loop
+    if(code[[1]] == 'for') {
+        code[[4]] <- codeProcessModelMacros(code[[4]], envir)
+        return(code)
+    }
+    ## Check if this line invokes a submodel.
+    ## This can be done in two ways:
+    ## (i) node1 [<- | ~] <macro name>(...)
+    ## or
+    ## (ii) <macro name>(args)
+    ##
+    ## The first version is more BUGS-like.
+    ## The second version allows more full control.
+
+    ## Check for first version:
```

Comment:
This confused me a little, since at first glance I didn't see that both forms are supported. Maybe it would clearer to comment like
```r
possibleMacroName <- deparse(code[[1]])
## Reduce first form to second form.
if(possibleMacroName %in% c('<-', '~')) {
  possibleMacroName <- deparse(code[[3]][[1]])
}
## Check for second form.
...
```